Implement missing functionality of AudioBufferSourceNode

collares opened the pull request:
This is most of the fix for [Servo's #22363](https://github.com/servo/servo/issues/22363). It is broken into four pieces. Let me know if I can help in the reviewing process.

ferjm commented on audio/buffer_source_node.rs:
```diff
@@ -61,12 +70,19 @@ pub(crate) struct AudioBufferSourceNode {
+    start_duration: Option<f64>,
+    /// The same as start_at, but with subsample accuracy.
+    /// FIXME: Maybe AudioScheduledSourceNode should use this as well?
+    start_when: f64,
```
Yes, please. Let's make this part of `AudioScheduledSourceNode`. It could be done in a follow-up though.

ferjm commented on audio/buffer_source_node.rs:
```diff
@@ -8,6 +8,8 @@ use param::{Param, ParamType};
     /// Set the data block holding the audio sample data to be played.
     SetBuffer(Option<AudioBuffer>),
+    /// Set start parameters (when, offset, duration).
+    Start(f64, Option<f64>, Option<f64>),
```
nit: rename this to `SetStartParams` or such, so it is not confused with the `AudioScheduledSourceNodeMessage::Start` message.

It would also be nice to remove the first parameter (`when`) as we can get it from `AudioScheduledSourceNodeMessage::Start`.

collares replied:
I fixed the first part, and will leave the second part to the AudioScheduledSourceNode follow-up.

collares commented:
Addressed review comments! There are two remaining relevant WPT failures:

1) I have no idea what's happening with sub-sample-buffer-stitching.html. If I force the test to run, we seem to be huge values at the end of the test (of the order of 1e20), but the test only contanis samples in the range [-1, 1]. I don't see how any part of my code could generate such large values.

2) buffer-resampling.html fails because we interpolate the last sample with 0 when we output a sample corresponding to buffer.len() - 1 < buffer_pos < buffer.len(). The test wants us to hold the last sample constant for such values of buffer_pos, but it is unclear to me if this is the right thing to do.

If you are OK with the above failures, this PR is ready to be committed.

ferjm commented:
Thanks! Could you file follow-up issues for the remaining work (i.e. all the XXX comments that you added to the code) so we don't lose track of it, please?

@bors-servo r+

bors-servo commented:
:pushpin: Commit 198f060 has been approved by `ferjm`

<!-- @bors-servo r=ferjm 198f0604e7355d80b25e1c3f2abfcf2177b93c2c -->

bors-servo commented:
:hourglass: Testing commit 198f0604e7355d80b25e1c3f2abfcf2177b93c2c with merge 86b9a2892af3774e42243ac84394a052719b2bf6...

bors-servo commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/servo/media/builds/123921297)
Approved by: ferjm
Pushing 86b9a2892af3774e42243ac84394a052719b2bf6 to master...
